### PR TITLE
[expo-cli] Add platform when running --dev-client

### PR DESCRIPTION
**Pull request by jonsamp**
# Why

When running `npx expo start --dev-client` when using modern manifests during development, the local development server returns:

```bash
{"error":"CommandError: Must specify \"expo-platform\" header or \"platform\" query parameter"}%
```

When running a local server, we cannot always be sure of the platform of the manifest we'll need, like when a developer is using Flipper. When a client would like to consume the manifest of a development server but is not an Android or iOS device, we now return "none" as the platform.

This problem is outlined in these issues: https://github.com/expo/expo/issues/17495, https://github.com/jakobo/expo-community-flipper/issues/24

# How

- Makes Expo CLI default to a platform of "none" when there is no platform specified.
- Adds a debug log stating this fall back if running `DEBUG=* npx expo start --dev-client`.
- Removes a platform assertion, since there will now always be a platform. Still checks for a valid platform.

# Test Plan

I made a repo to reproduce this issue: https://github.com/jonsamp/flipper-test. 

1. Clone that repo and install its dependencies.
2. Then, run `nexpo start --dev-client` (local Expo CLI on this branch)
3. Then, run `curl http://192.168.4.50:8081/` (with whichever server location is printed from starting the server)
4. You should see a manifest (as pictured below) return. You should not see `{"error":"CommandError: Must specify \"expo-platform\" header or \"platform\" query parameter"}`
   Screen Shot 2022-09-30 at 2 09 23 AM

### CODE REVIEW

**Comment by ide** [comment]
I mentioned this to Jon out of band as well but want to repeat it here: I don't think we should have a default platform since no value inherently makes sense as the default. (In the fullness of time, the one value I believe could qualify as a default is "universal" when we manage to get away from Android-, iOS-, or web-specific bundles and can produce one bundle that runs everywhere.)

Jon explained to me that the issue is Flipper (the debug tool) is requesting a manifest and Flipper doesn't run on Android nor iOS. In this case, I propose we add a dummy platform for this purpose, such as `expo-platform: none`. This means, "I'm not looking to run this on any device -- if I were, I'd need to make a request with `expo-platform: <device-platform>` instead."

**Comment by brentvatne** [comment]
@ide - I don't think we can go out and ask tools that try to connect to the react native dev server to add the expo-platform=none header to their request. If you are proposing that the absence of an expo-platform means we treat it as expo-platform=none, that seems fine.

**Comment by ide** [comment]
> If you are proposing that the absence of an expo-platform means we treat it as expo-platform=none, that seems fine.

I think this would also be acceptable. My underlying thinking is that Android devices should always load Android updates and iOS devices should always load iOS updates, or they should unambiguously fail. It's better to consistently fail rather than inconsistently succeed if a device is misconfigured and doesn't specify its platform. This is why I believe neither "ios" nor "android" should be the default and defaulting `expo-platform` to "none" satisfies this.

When a device, as opposed to a tool like Flipper, loads an update manifest with `expo-platform` set to "none", it'd be great to show a clear error message explaining:

- This Android/iOS device loaded an update for a different platform ("none"/"ios"/"android").
- This can happen when your device doesn't send the "expo-platform" HTTP header with its platform ("ios"/"android").
- Some tips for how to address this, if we have any.
- Bonus: a button to run the update for a different platform anyway, acknowledging it might not work

**Comment by jonsamp** [comment]
I tested the `"none"` platform and confirmed that both Flipper and an iOS simulator still work in development. Screenshot:

Screen Shot 2022-10-12 at 8 43 40 PM

---

I have two loose ends:
1. In Expo CLI, we manually define platforms as strings (like "android", or "ios"). I tried converting these into a typed `enum` to make sure that we only ever pass/use the correct platform, however, this added a lot of scope to this PR and involved changes to config-plugins. In the future, we might want to consider adding an `enum` for these platform strings.
2. If I manually navigate to "https://[my-server-location]:8081?platform=none" inside of a development build, I get an error stating "Unable to resolve module ./Libraries/Components/DatePickerIOS from ...". This is a vague and strange error, where the cause is that the platform is incorrect. As a follow up task, where can we catch that error and display the message @ide [suggested](https://github.com/expo/expo/pull/19319#issuecomment-1265968763).
   <details><summary>Screenshot of error</summary> Screen Shot 2022-10-12 at 8 47 47 PM  </details>

File: packages/@expo/cli/src/start/server/middleware/resolvePlatform.ts [hunk 1]
```diff
@@ -52,7 +52,7 @@ export function assertMissingRuntimePlatform(platform?: any): asserts platform {
 /** Assert if the runtime platform is not correct. */
 export function assertRuntimePlatform(platform: string): asserts platform is RuntimePlatform {
   const stringifiedPlatform = String(platform);
-  if (!['android', 'ios'].includes(stringifiedPlatform)) {
+  if (!['android', 'ios', 'none'].includes(stringifiedPlatform)) {
     throw new CommandError(
       'PLATFORM_HEADER',
       `platform must be "android" or "ios". Received: "${platform}"`
```
Comment (ide) [inline packages/@expo/cli/src/start/server/middleware/resolvePlatform.ts:58]: Update this error message now that "none" is a valid value.

**Comment by ide** [comment]
For (2):

- The manifest for "none" could just have `data:,` or `data:application/javascript,` (empty data URIs) as the launch asset URL. There is no launch asset (after all it's an update for no platform) and an empty data URI still lets us make sure there's a launch asset.
- The dev server endpoint for bundles probably should check for `platform=none` and special-case the response either to be a 400 error (if we consider it to be illogical to ask for a bundle for no platform) or a 204 No Content response, which matches the empty data URI. Open to other suggestions as long as they allow properly configured projects to work and improperly configured projects not to work.

**Review by ide — APPROVED** [review]

**Comment by ide** [comment]
CI unit test failure looks legit and easy to fix.

**Review by brentvatne — APPROVED** [review]

**Pull request merged**